source/grid/tria_description.cc

```diff
@@ -0,0 +1,662 @@
+
+                //    store the coarse cell id
+                cell_info.id = CellId(new_coarse_cell_id, 0, nullptr)
+                                 .template to_binary<dim>();
+
+                //    save coarse_cell_index -> coarse_cell_id mapping
+                construction_data.coarse_cell_index_to_coarse_cell_id.push_back(
+                  new_coarse_cell_id);
+
+                // e) store manifold id of cell
+                cell_info.manifold_id = cell->manifold_id();
+
+                // ... of lines
+                if (dim >= 2)
+                  for (unsigned int line = 0;
+                       line < GeometryInfo<dim>::lines_per_cell;
+                       ++line)
+                    cell_info.manifold_line_ids[line] =
+                      cell->line(line)->manifold_id();
+
+                // ... of quads
+                if (dim == 3)
+                  for (unsigned int quad = 0;
+                       quad < GeometryInfo<dim>::quads_per_cell;
+                       ++quad)
+                    cell_info.manifold_quad_ids[quad] =
+                      cell->quad(quad)->manifold_id();
+
+                // f) store subdomain_id
+                cell_info.subdomain_id = cell->subdomain_id();
+
+                // g) store invalid level_subdomain_id (since multilevel
+                //    hierarchy is not constructed)
+                cell_info.level_subdomain_id = numbers::invalid_subdomain_id;
+
+                construction_data.cell_infos[0].emplace_back(cell_info);
+              }
+
+          // 3) enumerate locally relevant vertices
+          unsigned int vertex_counter = 0;
+          for (auto &vertex : vertices_locally_relevant)
+            {
+              construction_data.coarse_cell_vertices.push_back(
+                tria.get_vertices()[vertex.first]);
+              vertex.second = vertex_counter++;
+            }
+
+          // 4) correct vertices of cells (make them local)
+          for (auto &cell : construction_data.coarse_cells)
+            for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+                 ++v)
+              cell.vertices[v] = vertices_locally_relevant[cell.vertices[v]];
+        }
+      else
+        {
+          construction_data.settings =
+            TriangulationDescription::Settings::construct_multigrid_hierarchy;
+
+          // 1) collect locally relevant cells (set user_flag)
+          std::vector<bool> old_user_flags;
+          tria.save_user_flags(old_user_flags);
+
+          // 1a) clear user_flags
+          const_cast<dealii::Triangulation<dim, spacedim> &>(tria)
+            .clear_user_flags();
+
+          // 1b) loop over levels (from fine to coarse) and mark on each level
+          //     the locally relevant cells
+          for (unsigned int level =
+                 tria.get_triangulation().n_global_levels() - 1;
+               level != numbers::invalid_unsigned_int;
```

Comment:
This uses special knowledge of what `invalid_unsigned_int` actually is. We tend to think of this as just some exceptional value like a `NaN` that lives outside the usual values. Just make `level` a `signed int` and use this as condition:
```suggestion
               level >= 0;
```